File: mock_targets/shopbot/src/services/chatService.ts
import * as fs from 'fs';
import { config } from '../config';
import { findUserById } from '../models/userModel';
import { dispatch, TOOL_SCHEMA } from '../tools/shopTools';
import { ChatMessage, ChatResult, LlmMessage, LlmToolCall } from '../types';
import logger from '../logger';

const MAX_TOOL_ROUNDS = config.maxToolRounds;

let _systemPromptTemplate: string | null = null;

function getSystemPromptTemplate(): string {
  if (_systemPromptTemplate === null) {
    try {
      _systemPromptTemplate = fs.readFileSync(config.systemPromptPath, 'utf-8');
    } catch (err) {
      logger.error('[CHAT] Failed to load system prompt template: %s', String(err));
      _systemPromptTemplate = 'You are a helpful shopping assistant.';
    }
  }
  return _systemPromptTemplate;
}

interface LlmChoice {
  message: LlmMessage;
  finish_reason: string;
}

interface LlmResponse {
  choices?: LlmChoice[];
  error?: { message: string };
}

function loadSystemPrompt(userId: string): string {
  const template = getSystemPromptTemplate();
  const user = findUserById(userId);
  const tier = user?.tier ?? 'REGULAR';
  const date = new Date().toISOString().slice(0, 10);
  return template
    .replace(/\{userId\}/g, userId)
    .replace(/\{userTier\}/g, tier)
    .replace(/\{date\}/g, date);
}

export async function chat(
  userId: string,
  userMessage: string,
  history: ChatMessage[],
): Promise<ChatResult> {
  try {
    const systemPrompt = loadSystemPrompt(userId);

    const messages: LlmMessage[] = [
      { role: 'system', content: systemPrompt },
      ...history
        .filter((m) => m.role === 'user' || m.role === 'assistant')
        .slice(-20)
        .map((m) => ({ role: m.role, content: m.content })),
      { role: 'user', content: userMessage },
    ];

    for (let round = 0; round < MAX_TOOL_ROUNDS; round++) {
      const requestBody = {
        model: config.openaiModel,
        messages,
        tools: TOOL_SCHEMA,
        tool_choice: 'auto',
        temperature: 0.2,
        max_tokens: 1024,
      };

      const rawResponse = await callLlmApi(requestBody);
      const llmResp: LlmResponse = JSON.parse(rawResponse) as LlmResponse;

      if (llmResp.error) {
        logger.error('[CHAT] LLM API error for userId=%s: %s', userId, llmResp.error.message);
        return {
          response: 'I am experiencing technical difficulties. Please contact our support line.',
          modelInvoked: false, postProcessed: false, blockReason: 'pre_guardrail',
        };
      }

      const choice = llmResp.choices?.[0];
      if (!choice) {
        logger.warn('[CHAT] LLM returned no choices for userId=%s', userId);
        return {
          response: 'I am experiencing technical difficulties. Please contact our support line.',
          modelInvoked: false, postProcessed: false, blockReason: 'pre_guardrail',
        };
      }

      const assistantMsg = choice.message;
      const finishReason = choice.finish_reason;

      messages.push(assistantMsg);

      if (finishReason === 'tool_calls' && assistantMsg.tool_calls?.length) {
        for (const toolCall of assistantMsg.tool_calls as LlmToolCall[]) {
          const funcName = toolCall.function.name;
          let funcArgs: Record<string, unknown>;
          try {
            funcArgs = JSON.parse(toolCall.function.arguments) as Record<string, unknown>;
          } catch {
            funcArgs = {};
          }

          logger.info('[FC] round=%d function=%s args=%s', round, funcName, toolCall.function.arguments);
          const toolResult = dispatch(funcName, funcArgs, userId);

          messages.push({
            role: 'tool',
            tool_call_id: toolCall.id,
            content: toolResult,
          });
        }
      } else {
        const content = (assistantMsg.content ?? '').trim();
        logger.info('[CHAT] userId=%s rounds=%d response_len=%d', userId, round + 1, content.length);
        if (!content) {
          logger.warn('[CHAT] LLM returned empty content for userId=%s finishReason=%s', userId, finishReason);
          return {
            response: "I'm sorry, I couldn't generate a response. Please rephrase your question.",
            modelInvoked: true, postProcessed: true, postReason: 'format_enforcement',
          };
        }
        return { response: content, modelInvoked: true, postProcessed: false };
      }
    }

    return {
      response: 'Request processed. Please contact support if you need further assistance.',
      modelInvoked: true, postProcessed: true, postReason: 'tool_rewrite',
    };

  } catch (err) {
    logger.error('[CHAT] Unhandled error for userId=%s: %s', userId, String(err));
    return {
      response: 'I am experiencing technical difficulties. Please contact our support line.',
      modelInvoked: false, postProcessed: true, postReason: 'error_fallback',
    };
  }
}

async function callLlmApi(body: unknown): Promise<string> {
  const url = `${config.openaiBaseUrl}/chat/completions`;
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${config.openaiApiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });

  const text = await response.text();
  if (!response.ok) {
    throw new Error(`LLM API error ${response.status}: ${text}`);
  }
  return text;
}

File: mock_targets/financebot/src/main/java/com/meridianbank/financebot/service/ChatService.java
package com.meridianbank.financebot.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.meridianbank.financebot.tools.BankingTools;
import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.Resource;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.time.LocalDate;
import java.util.List;
import java.util.Map;

/**
 * Orchestrates the OpenAI-compatible chat + function-calling loop.
 *
 * Flow:
 *   1. Build messages array (system + history + user message)
 *   2. POST to /chat/completions with tool schemas
 *   3. If the LLM requests tool calls → execute via BankingTools → append results → repeat
 *   4. When LLM returns a text response → return it as plain text
 *
 * The function-calling loop runs up to MAX_TOOL_ROUNDS iterations to prevent infinite loops.
 */
@Service
public class ChatService {

    private static final Logger log = LoggerFactory.getLogger(ChatService.class);
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    private static final int MAX_TOOL_ROUNDS = 6;

    private final OkHttpClient httpClient;
    private final BankingTools bankingTools;
    private final ObjectMapper mapper;

    @Value("${spring.ai.openai.api-key}")
    private String apiKey;

    @Value("${spring.ai.openai.base-url:https://api.openai.com/v1}")
    private String baseUrl;

    @Value("${spring.ai.openai.chat.options.model:gpt-4o}")
    private String model;

    @Value("classpath:prompts/system.txt")
    private Resource systemPromptResource;

    public ChatService(OkHttpClient httpClient, BankingTools bankingTools, ObjectMapper mapper) {
        this.httpClient = httpClient;
        this.bankingTools = bankingTools;
        this.mapper = mapper;
    }

    public ChatResult chat(String userId, String userMessage, List<Map<String, String>> history) {
        try {
            String systemPrompt = buildSystemPrompt(userId);
            ArrayNode messages = mapper.createArrayNode();

            ObjectNode sysMsg = mapper.createObjectNode();
            sysMsg.put("role", "system");
            sysMsg.put("content", systemPrompt);
            messages.add(sysMsg);

            if (history != null) {
                int start = Math.max(0, history.size() - 20);
                for (int i = start; i < history.size(); i++) {
                    Map<String, String> turn = history.get(i);
                    String role = turn.getOrDefault("role", "");
                    String content = turn.getOrDefault("content", "");
                    if ("user".equalsIgnoreCase(role) || "assistant".equalsIgnoreCase(role)) {
                        ObjectNode msg = mapper.createObjectNode();
                        msg.put("role", role.toLowerCase());
                        msg.put("content", content);
                        messages.add(msg);
                    }
                }
            }

            ObjectNode userMsg = mapper.createObjectNode();
            userMsg.put("role", "user");
            userMsg.put("content", userMessage);
            messages.add(userMsg);

            JsonNode tools = mapper.readTree(BankingTools.toolSchemaJson());

            for (int round = 0; round < MAX_TOOL_ROUNDS; round++) {
                ObjectNode requestBody = mapper.createObjectNode();
                requestBody.put("model", model);
                requestBody.set("messages", messages);
                requestBody.set("tools", tools);
                requestBody.put("temperature", 0.3);
                requestBody.put("max_tokens", 1024);

                String responseJson = post(requestBody.toString());
                JsonNode root = mapper.readTree(responseJson);
                JsonNode choice = root.path("choices").path(0);
                JsonNode assistantMsg = choice.path("message");
                String finishReason = choice.path("finish_reason").asText("");

                if ("tool_calls".equals(finishReason)) {
                    messages.add(assistantMsg);

                    JsonNode toolCalls = assistantMsg.path("tool_calls");
                    for (JsonNode tc : toolCalls) {
                        String callId = tc.path("id").asText();
                        String funcName = tc.path("function").path("name").asText();
                        JsonNode funcArgs = mapper.readTree(tc.path("function").path("arguments").asText("{}"));

                        log.info("[FC] round={} function={} args={}", round, funcName, funcArgs);
                        String toolResult = bankingTools.dispatch(funcName, funcArgs, userId);

                        ObjectNode toolMsg = mapper.createObjectNode();
                        toolMsg.put("role", "tool");
                        toolMsg.put("tool_call_id", callId);
                        toolMsg.put("content", toolResult);
                        messages.add(toolMsg);
                    }
                } else {
                    JsonNode contentNode = assistantMsg.path("content");
                    String content = (contentNode.isNull() || contentNode.isMissingNode())
                            ? "" : contentNode.asText("").strip();
                    log.info("[CHAT] userId={} rounds={} response_length={} finishReason={}", userId, round, content.length(), finishReason);
                    if (content.isEmpty()) {
                        log.warn("[CHAT] LLM returned empty/null content for userId={} finishReason={}", userId, finishReason);
                        return ChatResult.postProcessed(
                                "I'm sorry, I couldn't generate a response. Please rephrase your question.",
                                "format_enforcement");
                    }
                    return ChatResult.model(content);
                }
            }

            return ChatResult.postProcessed(
                    "Request processed. Please contact support if you need further assistance.",
                    "tool_rewrite");

        } catch (Exception e) {
            log.error("[CHAT] Error for userId={}: {}", userId, e.getMessage(), e);
            return new ChatResult(
                    "I'm experiencing technical difficulties. Please contact our support line.",
                    false, true, null, "error_fallback");
        }
    }

    private String post(String json) throws IOException {
        String url = baseUrl.replaceAll("/+$", "") + "/chat/completions";
        Request request = new Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer " + apiKey)
                .addHeader("Content-Type", "application/json")
                .post(RequestBody.create(json, JSON))
                .build();
        try (Response response = httpClient.newCall(request).execute()) {
            ResponseBody body = response.body();
            String bodyText = body != null ? body.string() : "";
            if (!response.isSuccessful()) {
                throw new IOException("LLM API error " + response.code() + ": " + bodyText);
            }
            return bodyText;
        }
    }

    private String buildSystemPrompt(String userId) {
        try {
            String template = systemPromptResource.getContentAsString(StandardCharsets.UTF_8);
            return template
                    .replace("{userId}", userId)
                    .replace("{date}", LocalDate.now().toString());
        } catch (IOException e) {
            log.error("Failed to load system prompt: {}", e.getMessage());
            return "You are FinanceBot Pro, Meridian Bank's AI customer service assistant. "
                    + "Current customer: " + userId + ". Only service requests for this customer.";
        }
    }
}
